packages/effect/src/internal/stream.ts

```diff
@@ -4686,6 +4686,44 @@ export const range = (min: number, max: number, chunkSize = DefaultChunkSize): S
     return new StreamImpl(go(min, max, chunkSize))
   })
 
+export const raceAll = <S extends ReadonlyArray<Stream.Stream<any, any, any>>>(
+  ...streams: S
+): Stream.Stream<
+  Stream.Stream.Success<S[number]>,
+  Stream.Stream.Error<S[number]>,
+  Stream.Stream.Context<S[number]>
+> =>
+  Deferred.make<void>().pipe(
+    Effect.map((halt) => {
+      let winner: number | null = null;
+      return mergeAll(
+        streams.map((stream, index) =>
+          stream.pipe(
+            takeWhile(() => {
+              if (winner === index) {
+                return true;
+              } else if (winner === null) {
+                winner = index;
+                Deferred.unsafeDone(halt, Exit.void);
+                return true;
+              }
+              return false;
+            }),
```

Comment:
This feels a bit cleaner

```suggestion
            takeWhile(() => {
              if (winner === null) {
                winner = index;
                Deferred.unsafeDone(halt, Exit.void);
                return true;
              }
              return winner === index;
            }),
```

Reply:
Yes, i thought about it, but while it really feels cleaner, i decided to go that way because of perfomance.

The predicate function will be executed only once for all the loosers, but can be executed an infinite amount of times for the winner. 

It means that if i apply your suggestion, each item will be checked to `null` first, and then checked to `winner === index`, so it is 2 checks per each emission. 

I optimised it for the winner stream by placing the check for `winner === index` first, so the first check is successful for the winner branch, which is the most sensitive branch than the loosers.

Reply:
The performance impact here will be zero, so I generally lead towards readability. But I'm OK with either approach.